keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,138 @@
+---
+kep-number: 27
+short-desc: A parameter flag to disable forced pod reload
+title: Forced Pod reload attribute
+authors:
+  - @aneumann82
+owners:
+  - @aneumann82
+editor:
+creation-date: 2020-03-31
+last-updated: 2020-03-31
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+## Summary
+
+This KEP describes addition of a flag for parameters that controls the forced pod reloading
+
+## Motivation
+
+The new flag will allow an operator developer to control that a specific attribute does not automatically restart
+all pods of a deployment or stateful set if this parameter is changed.
+
+### Goals
+
+Make it possible for an operator developer to specify that a parameter will *not* automatically restart all pods from
+a stateful set or a deployment.
+
+### Non-Goals
+
+- Detailed control when Pods are restarted: If the spec template for a stateful set is changed, Kubernetes will still
+reload all pods automatically.
+- Calculation of used config maps, secrets or other resources used by a pod: The automatic pod restart is currently 
+used to force pods to re-read config maps on a parameter update. When a user explicitly disables the pod restart for 
+a parameter, and this parameter updates a config map or another dependency of the pod, the change might not be applied
+in the pod until the next restart.
+
+## Proposal
+
+Add an additional attribute `forcePodRestart` to parameter specifications in `params.yaml`:
+
+```yaml
+  - name: NODE_COUNT
+    description: "Number of Cassandra nodes."
+    default: "3"
+    forcePodRestart: "false"
```

Comment:
I'm wondering: is this something we need per parameter? Or per `kudo update` command execution? Can the same parameter either update pod templates or not based on the specific use case? And if it is based on the use case, can we solve this by adding another option to `kudo update ... --restartPods`?

Reply:
I think this could be solved with `kudo update --restartPods` as well, yes. The flag per parameter basically automates the setting of the flag: If one of the modified parameters has the flag set to true, the `--restartPods` is set.

I would prefer having it on the parameters, as it's then only the responsibility of the operator developer to think about when to apply these settings, not the human operator user - who would need to think about the flag every time they update a parameter.

It might still be a nice idea to add the `--restartPods` parameter to override the behavior. Not sure if it's really needed though